File: medtronic/src/main/java/info/nightscout/androidaps/plugins/pump/common/data/PumpStatus.java
package info.nightscout.androidaps.plugins.pump.common.data;

import java.util.Date;

import info.nightscout.androidaps.interfaces.ProfileStore;
import info.nightscout.androidaps.plugins.pump.common.defs.PumpStatusType;
import info.nightscout.androidaps.plugins.pump.common.defs.PumpType;
import info.nightscout.androidaps.utils.DateUtil;

/**
 * Created by andy on 4/28/18.
 */

public abstract class PumpStatus {

    // connection
    public long lastDataTime;
    public long lastConnection = 0L;
    public long previousConnection = 0L; // here should be stored last connection of previous session (so needs to be
    // read before lastConnection is modified for first time).

    // last bolus
    public Date lastBolusTime;
    public Double lastBolusAmount;

    // other pump settings
    public String activeProfileName = "0";
    public double reservoirRemainingUnits = 0.0d;
    public int reservoirFullUnits = 0;
    public int batteryRemaining = 0; // percent, so 0-100
    public Double batteryVoltage = null;


    // iob
    public String iob = null;

    // TDD
    public Double dailyTotalUnits;
    public String maxDailyTotalUnits;
    public boolean validBasalRateProfileSelectedOnPump = true;
    public ProfileStore profileStore;
    public String units; // Constants.MGDL or Constants.MMOL
    public PumpStatusType pumpStatusType = PumpStatusType.Running;
    public Double[] basalsByHour;
    public double currentBasal = 0;
    public int tempBasalInProgress = 0;
    public int tempBasalRatio = 0;
    public int tempBasalRemainMin = 0;
    public Date tempBasalStart;
    public PumpType pumpType;
    //protected PumpDescription pumpDescription;


    public PumpStatus(PumpType pumpType) {
        //  public PumpStatus(PumpDescription pumpDescription) {
        //       this.pumpDescription = pumpDescription;

//        this.initSettings();
        this.pumpType = pumpType;
    }

    public abstract void initSettings();

    public void setLastCommunicationToNow() {
        this.lastDataTime = DateUtil.now();
        this.lastConnection = System.currentTimeMillis();
    }

    public abstract String getErrorInfo();

}

File: pump/pump-common/src/main/java/info/nightscout/aaps/pump/common/data/PumpStatus.kt
package info.nightscout.aaps.pump.common.data

import info.nightscout.aaps.pump.common.data.PumpTimeDifferenceDto
import info.nightscout.interfaces.pump.defs.PumpType
import info.nightscout.aaps.pump.common.defs.PumpRunningState
import info.nightscout.aaps.pump.common.defs.TempBasalPair
import java.util.Date

/**
 * Created by andy on 4/28/18.
 */
abstract class PumpStatus(var pumpType: PumpType) {

    // connection
    var lastDataTime: Long = 0
    var lastConnection = 0L
    var previousConnection = 0L // here should be stored last connection of previous session (so needs to be

    // bolus
    var lastBolusTime: Date? = null
    var lastBolusAmount: Double? = null

    // other pump settings
    var reservoirRemainingUnits = 0.0
    var reservoirFullUnits = 0
    var batteryRemaining = 0 // percent, so 0-100
    var batteryVoltage: Double? = null
    var units: String? = null // Constants.MGDL or Constants.MMOL

    // iob
    var iob: String? = null

    // basal profile
    var basalsByHour: DoubleArray? = null
    var activeProfileName = "1"

    // TDD
    var dailyTotalUnits: Double? = null
    var maxDailyTotalUnits: String? = null


    // state
    var pumpRunningState = PumpRunningState.Running

    // temp basal
    var currentTempBasal: TempBasalPair? = null
    var currentTempBasalEstimatedEnd: Long? = null

    // time
    var pumpTime: PumpTimeDifferenceDto? = null


    // TODO refactor to use TempBasalPair - remove this
    var tempBasalStart: Long? = null
    var tempBasalAmount: Double? = 0.0
    var tempBasalPercent: Int? = 100
    var tempBasalDuration: Int? = 0
    var tempBasalEnd: Long? = null


// OLD - Start
//     var units: String? = null // Constants.MGDL or Constants.MMOL
//     var pumpRunningState = PumpRunningState.Running
//     var basalsByHour: DoubleArray? = null
//     var tempBasalStart: Long? = null
//     var tempBasalAmount: Double? = 0.0
//     var tempBasalLength: Int? = 0
//     var tempBasalEnd: Long? = null
//     var pumpTime: PumpTimeDifferenceDto? = null
// OLD - End

    abstract fun initSettings()

    fun setLastCommunicationToNow() {
        lastDataTime = System.currentTimeMillis()
        lastConnection = System.currentTimeMillis()
    }

    abstract val errorInfo: String?

}
